internal/indexer/indexer.go

```diff
@@ -854,6 +866,497 @@ func (i *Indexer) indexReferenceToExternalDefinition(p *packages.Package, docume
 	return rangeID, true
 }
 
+type def struct {
+	pkg   *packages.Package
+	obj   types.Object
+	ident *ast.Ident
+}
+
+func extractTypes(pkgs []*packages.Package) ([]def, []def) {
+	interfaces := []def{}
+	concreteTypes := []def{}
+	for _, pkg := range pkgs {
+		for ident, obj := range pkg.TypesInfo.Defs {
+			if obj == nil {
+				continue
+			}
+
+			// We ignore aliases 'type M = N' to avoid duplicate reporting
+			// of the Named type N.
+			if _, ok := obj.(*types.TypeName); !ok {
+				continue
+			}
+
+			if _, ok := obj.Type().(*types.Named); !ok {
+				continue
+			}
+
+			// methodLen := types.NewMethodSet(obj.Type()).Len()
+			// methodLen := listMethods(obj.Type())
+			methodLen := len(listMethods(obj.Type().(*types.Named)))
+			if methodLen == 0 {
+				// fmt.Println("OH NO NO NO", obj.Name())
+				continue
+			}
+
+			if types.IsInterface(obj.Type()) {
+				// TODO figure out non-exported interfaces
+				// should link within package? across packages?
+				interfaces = append(interfaces, def{pkg: pkg, obj: obj, ident: ident})
+			} else {
+				concreteTypes = append(concreteTypes, def{pkg: pkg, obj: obj, ident: ident})
+			}
+		}
+	}
+
+	return interfaces, concreteTypes
+}
+
+// indexImplementations emits data for each implementation of an interface.
+func (i *Indexer) indexImplementations() error {
+	// Load all dependencies
+	deps, err := i.loadDependencyPackages()
+	if err != nil {
+		return err
+	}
+
+	start1 := time.Now()
+	localInterfaces, localConcreteTypes := extractTypes(i.packages)
+	remoteInterfaces, remoteConcreteTypes := extractTypes(deps)
+	fmt.Println("## Extract Types:", time.Since(start1), "<==")
+
+	fmt.Println("localInterfaces", len(localInterfaces))
+	fmt.Println("localConcreteTypes", len(localConcreteTypes))
+	fmt.Println("remoteInterfaces", len(remoteInterfaces))
+	fmt.Println("remoteConcreteTypes", len(remoteConcreteTypes))
+
+	// TODO prune interfaces/types that are not exported
+	// TODO prune interfaces/types by method count
+
+	doTheRealThing := true
+	if doTheRealThing {
+		const local = 0
+		const remote = 1
+		const invertNo = false
+		const invertYes = true
+
+		positionToResultSet := map[string]uint64{}
+		ensureResultSet := func(d def) uint64 {
+			position, _, _ := i.positionAndDocument(d.pkg, d.ident.Pos())
+			if v, ok := positionToResultSet[position.String()]; ok {
+				return v
+			}
+			defInfo := i.getDefinitionInfo(d.obj, d.ident)
+			positionToResultSet[position.String()] = defInfo.ResultSetID
+			return defInfo.ResultSetID
+		}
+
+		monikerKeyToID := map[string]uint64{}
+
+		emitImplementations := func(concreteTypes, interfaces []def, rightKind int, shouldInvert bool) {
+			relation := i.buildImplementationRelation(concreteTypes, interfaces)
+			leftDefs, rightDefs := concreteTypes, interfaces
+
+			if shouldInvert {
+				relation = invert(relation)
+				leftDefs, rightDefs = rightDefs, leftDefs
+			}
```

Comment:
I have been wondering a bit about whether we should teach the backend that this language is reciprocal or whether we should actually emit both ways.

For example, if we say "struct X implements interface Y", is it always true that "interface Y is implemented by X"?

If so, then we do not need to emit everything for both directions. We can simply have:

```
X --textDocument/implement --> implementationResult --item--> Y
```

And the backend must handle the face that the reverse relationship is true as well.

cc @efritz so we can chat about this.

Reply:
The relationship _should_ be symmetric (imo, just by intuition). I was chatting with @chrismwendt about this and I think it might have some tendrils in the fact that we need to emit relationships for both:

- interfaces defined in a package and a conforming struct defined in a dependency, and
- structs defined in a package and a descriptive interface of that struct defined in a dependency

Reply:
Are defs/refs symmetric, too? (I think so) If so, should indexers only emit one direction and the backend fill in the other? If not, there are probably some practical or principled reason(s) why that can help us decide whether or not implementations should follow the same model.

This question applies to implementation relationships within an LSIF dump, but not across dumps because cross-dump find-implementations uses monikers rather than `textDocument/implementation` et al.